### DISCUSSION TITLE
Streamline and update documentation

### DISCUSSION THREAD

**Discussion by Helldembez** [etlegacy/etlegacy#1776] [General]
For a while now it's been going through my head that documentation needs to be updated. Even as I was making features/bug fixes I thought about it. But when thinking about it it feels like "where do I start?", since it's behind by quite a bit.

My suggestion is to think about this together and break up this thing what feels like a mountain into manageable pieces. Also to prevent the mountain getting bigger we should integrate this into the review process. Before merging or approving a PR, think if there is anything to be updated in the documentation.(cvars, intended behaviour etc) I feel like the review process should also be described on the wiki just like https://github.com/etlegacy/etlegacy/wiki/How-to-commit-Your-Code for example.

Which brings me to another observation I made. Merging straight to master. I understand the efficiency of it, but I feel like some things get pushed that should be reviewed. Maybe that already happens 1 on 1 between devs, but just a point of attention. Of coarse still push things like typo's or anything that's on that level.

I'll start by saying that I don't have an idea how to figure out documentation of cvars except for reading and debugging the code. This feels a bit inefficient I guess. But as a small task, I guess start by updating the cvars documentation you personally introduced or updated. What do you think?

**Comment by rmarquis** [comment]
You're right that the documentation is kinda mess: we have the [wiki](https://github.com/etlegacy/etlegacy/wiki), the general [documentation](https://etlegacy.readthedocs.io/) (wip), and the [Lua API doc](https://etlegacy-lua-docs.readthedocs.io/), and probably some other stuff I forgot about.

*But* the state of that documentation is in much better shape than some years ago, and the reason it is not in an even better shape yet is that it is an enormous effort to undertake. It's not a new topic either, we have had many people that wanted to make it better over time but quickly gave up when they realized what mess they put themselves into (and I don't blame them).

The only doc I can confidently say is "okayish" is the Lua API doc, because I put myself quite some hours to update it (basically took the etpro doc as a base, checked it entirely with the existing derived docs of some mods, and updated it with what we added over time). And that's a small part only of the general documentation topic.

There is also the wiki vs readthedoc format question: The latter is nicer to look at, can be printed/exported to PDF, but requires Git to update and maintain. The wiki is easier to manage to most contributors, and information that is often modified is imho better put into a wiki.

This said, I'm all for a well done, well maintained and complete cvars doc, but for this we need someone to put the hours into - and it is unfortunately for the most part boring work. Some work has been done already, but I feel we really need a dedicated person with lot of available time to ensure an overall good quality, as opposed to many little helping hands that do help, but might not have the general overview to achieve that project.

Also, another possibility might be to document cvars in the code directly (as opposed to do it manually), and somehow export it to a more readable format. Something like Javadoc, there's plenty of software that do something similar, maybe we can find something that fits our need here.

(I'll continue the discussion about review a it later when I got time).

**Reply by Helldembez** [reply]
BECAUSE its an enormous effort to undertake I'm convinced to break it up in small edible pieces. Just work step by step. It doesn't have to be done tomorrow. Maybe one person can start with cg_ cvars for example. or someone else picks up the task of adding hud customization tutorial(i want to do this already)

I was thinking about generating from code as well. Maybe we can use [doxygen](https://www.doxygen.nl/index.html). There even is a doxyfile already :)

**Reply by rmarquis** [reply]
Yes, but again: you're not the first person to want to improve the state of documentation. You're basically the last in a long line of fallen soldiers here (well, not that long, but you get the idea). What we need is continuous, sustained effort over a rather long time to improve the situation, and someone to lift the majority of the effort. Feel free to undertake this, if you feel like making your life goal on the shorter/medium term :)

Btw: some part of it have been done already, see the doc link above.

**Reply by rmarquis** [reply]
> Also, another possibility might be to document cvars in the code directly (as opposed to do it manually), and somehow export it to a more readable format.

I just realized that we have some built-in documentation stuff that isn't used, but that is extensively used by [ETe](https://github.com/etfdevs/ETe). There is a `Cvar_SetDescription` command that has been added in 2c3208f84e7d787c81643beeb745fcbb7e14b57e (see also #797), and there might be a way to dump the cvar list with their description.

**Comment by Aciz** [comment]
Having written the [ETJump documentation](https://etjump.readthedocs.io/en/latest/) myself basically from scratch, I know all too well what an effort it is, and a documentation for ETLegacy is an even bigger effort since we're talking about both mod and engine documentation.

However I must say, that for whatever weird reason, I quite like doing documentation, the issue is if I want to allocate my time to fully do it myself. I think splitting the work among multiple people is definitely a wise thing to do, it's just so much work for any single person to do. The people doing it should also have some knowledge of the codebase in general, and of course the ability to read code and figure out what stuff does instead of just trail/error values for cvars, to make it as accurate as possible. This is especially true for some of the more obsucre stuff, such as some renderer cvars (who can tell me what `r_railSegmentLength` does? yeah...)

We should also define a clear style and structure for stuff such as cvar documentation, if multiple people work with it, to avoid possibly having to rewrite a lot of things after the fact once you notice that someone has formatted things completely different.

**Reply by Helldembez** [reply]
If people are up for it we could also introduce some pair-writing through voice/screenshare when starting on something to share the knowledge and approach. This might incentivize people and speed up things :)

+1 for defining a clear style and structure. Even if only one person were to do it this is needed for future contributions. As I mentioned this should be defined on the wiki as well.
